client/client.go

```diff
@@ -191,17 +191,31 @@ func (c *httpClusterClient) Do(ctx context.Context, act httpAction) (*http.Respo
 		return nil, nil, errors.New("unable to pick endpoint: copy failed")
 	}
 
-	var resp *http.Response
-	var body []byte
-	var err error
+	var (
+		resp     *http.Response
+		body     []byte
+		err      error
+		epctx    context.Context
+		epcancel context.CancelFunc
+	)
 
 	for _, ep := range eps {
+		deadline, ok := ctx.Deadline()
+		if ok {
+			epctx, epcancel = context.WithTimeout(ctx, time.Duration(int(deadline.Sub(time.Now()))/len(eps)))
+		} else {
+			epctx = epctx
```

Comment:
yes... this shit is not tested... want to know if this is the approach you like.